File: tests/terminal-marker-fail-closed.test.ts
import { spawnSync } from "node:child_process";
import { randomBytes } from "node:crypto";
import { chmodSync, mkdirSync, mkdtempSync, rmSync, symlinkSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { entryExists } from "../src/cli/v2/store2";
import { handle } from "../src/browser/engine/verbs";
import { MemoryVfs, type Vfs } from "../src/browser/engine/vfs";
import type { EngineRequest, EngineResponse } from "../src/browser/engine/protocol";

/* ============================================================================
 * THE TERMINAL MARKER MUST FAIL CLOSED (FORMAT-V2.md §17).
 *
 * `destroyed.json` is the irreversible boundary: once it is durable the pair
 * must never perform a cryptographic operation again. The gate therefore asks
 * "is this path NOT KNOWN TO BE ABSENT", not "is there a readable regular file
 * here" — because every way of being unreadable must still close the boundary.
 *
 * A REAL FAIL-OPEN, MEASURED RATHER THAN ASSUMED. The gate used `existsSync`,
 * which follows symlinks and answers FALSE for a symlink whose target is gone.
 * A tombstone in that shape read as absent and the destroyed pair became usable
 * again — pad reuse, the one outcome TruePad may never allow. Node, the JVM and
 * Foundation all answer `false` there, which is why Android and iOS carried the
 * identical defect and were corrected in the same change. These are the
 * regression tests for the TypeScript half; the Kotlin and Swift halves are
 * TerminalMarkerFailClosedTest.kt and TerminalMarkerFailClosedTests.swift, and
 * all three assert the same list of shapes.
 * ========================================================================= */

const ROOT = resolve(__dirname, "..");
const LAUNCHER = join(ROOT, "bin", "truepad2.mjs");

let dir: string;
let sourceCount = 0;
beforeEach(() => {
  dir = mkdtempSync(join(tmpdir(), "truepad2-terminal-"));
});
afterEach(() => {
  rmSync(dir, { recursive: true, force: true });
});

function run(...argv: string[]): { code: number; stdout: string; stderr: string } {
  const child = spawnSync(process.execPath, [LAUNCHER, ...argv], { encoding: "utf8" });
  return { code: child.status ?? -1, stdout: child.stdout, stderr: child.stderr };
}

function genPair(pair: string, e = 64, n = 8): void {
  sourceCount += 1;
  const source = join(dir, `source-${sourceCount}.bin`);
  writeFileSync(source, randomBytes(2 * (e + 32 * n)));
  const gen = run("gen", pair, "--source", source, "--encryption-bytes", String(e), "--auth-records", String(n));
  expect(gen.code).toBe(0);
}

/** Every shape a tombstone can take that is NOT "a well-formed readable file". */
const SHAPES: ReadonlyArray<{ name: string; make: (path: string) => void }> = [
  { name: "a symlink whose target does not exist", make: (p) => symlinkSync(join(dir, "gone-target"), p) },
  {
    name: "a symlink to a deleted file",
    make: (p) => {
      const t = join(dir, "t.bin");
      writeFileSync(t, "x");
      symlinkSync(t, p);
      rmSync(t);
    }
  },
  { name: "a directory", make: (p) => mkdirSync(p) },
  {
    name: "a non-empty directory",
    make: (p) => {
      mkdirSync(p);
      writeFileSync(join(p, "inner"), "x");
    }
  },
  { name: "an empty file (a torn write)", make: (p) => writeFileSync(p, "") },
  { name: "a truncated JSON object", make: (p) => writeFileSync(p, '{"formatVersion":2,"pairId":"aaaa') },
  { name: "not JSON at all", make: (p) => writeFileSync(p, "  not json") },
  { name: "JSON that is not an object", make: (p) => writeFileSync(p, "[1,2,3]") },
  {
    name: "a JSON object naming a DIFFERENT pair",
    make: (p) => writeFileSync(p, JSON.stringify({ formatVersion: 2, pairId: "f".repeat(32) }))
  },
  {
    name: "a file with no read permission",
    make: (p) => {
      writeFileSync(p, "{}");
      chmodSync(p, 0);
    }
  }
];

describe("the §17 terminal marker fails closed in every shape", { timeout: 240_000 }, () => {
  for (const shape of SHAPES) {
    it(`refuses a consuming verb when destroyed.json is ${shape.name}`, () => {
      const a = join(dir, "a");
      genPair(a);

      // Control FIRST: without the marker this exact call succeeds, so the
      // refusal below is caused by the marker and nothing else.
      expect(run("burn", a, "--as", "A", "before").code).toBe(0);

      shape.make(join(a, "destroyed.json"));

      const burn = run("burn", a, "--as", "A", "after");
      expect(burn.code, `burn must refuse (${shape.name})`).toBe(2);
      expect(burn.stderr).toContain("refused: pair-destroyed");
      expect(burn.stdout).toBe("");

      const status = run("status", a);
      expect(status.code, `status must refuse (${shape.name})`).toBe(2);
      expect(status.stderr).toContain("refused: pair-destroyed");
    });
  }

  it("a destroyed pair whose tombstone is replaced by a dangling symlink STAYS destroyed", () => {
    // The end-to-end shape of the original defect: destroy for real, then let a
    // restore or a sync tool leave a broken link where the tombstone was.
    const a = join(dir, "a");
    genPair(a);
    const confirm = run("status", a).stdout.match(/[0-9a-f]{32}/)?.[0];
    expect(confirm).toBeDefined();
    expect(run("destroy", a, "--confirm", confirm as string).code).toBe(0);

    rmSync(join(a, "destroyed.json"));
    symlinkSync(join(dir, "gone-target"), join(a, "destroyed.json"));

    const burn = run("burn", a, "--as", "A", "resurrected?");
    expect(burn.code).toBe(2);
    expect(burn.stderr).toContain("refused: pair-destroyed");
  });

  it("entryExists reports presence for every non-absent shape, and absence only when definite", () => {
    for (const [index, shape] of SHAPES.entries()) {
      const p = join(dir, `probe-${index}`);
      shape.make(p);
      expect(entryExists(p), `${shape.name} must read as present`).toBe(true);
    }
    // The ONE definitive negative: there is no such path.
    expect(entryExists(join(dir, "nothing-here"))).toBe(false);
    expect(entryExists(join(dir, "no", "such", "directory", "at", "all.json"))).toBe(false);

    // A path whose parent is not a directory is NOT a definitive negative. Only
    // ENOENT is reported identically by Node, the JVM and Foundation, so it is
    // the only answer the three editions agree to treat as absence.
    const file = join(dir, "plain.bin");
    writeFileSync(file, "x");
    expect(entryExists(join(file, "under-a-file"))).toBe(true);
  });
});

/* ============================================================================
 * THE BROWSER HALF.
 *
 * OPFS has no symlinks, so the shape that broke the other three editions cannot
 * occur here — but the RULE still has to hold, and the way it can break in a
 * browser is different: `OpfsVfs.exists` resolves a file handle, and a directory
 * standing where a file belongs makes that call THROW `TypeMismatchError` rather
 * than return false. The gate must let that throw close the boundary; swallowing
 * it into "absent" would be the same fail-open in a different costume.
 * ========================================================================= */

let browserIdSeq = 1;
type WithoutId<T> = T extends { id: number } ? Omit<T, "id"> : never;
async function send(vfs: Vfs, req: WithoutId<EngineRequest>): Promise<EngineResponse> {
  return handle(vfs, { ...req, id: browserIdSeq++ } as EngineRequest);
}

/** Injects the failure mode a real OPFS backing produces for a non-file entry. */
class TypeMismatchOnExistsVfs implements Vfs {
  readonly inner: Vfs;
  readonly throwFor: (path: string) => boolean;
  constructor(inner: Vfs, throwFor: (path: string) => boolean) {
    this.inner = inner;
    this.throwFor = throwFor;
  }
  readFile(p: string) { return this.inner.readFile(p); }
  appendFile(p: string, d: Uint8Array) { return this.inner.appendFile(p, d); }
  writeFileAtomic(p: string, d: Uint8Array) { return this.inner.writeFileAtomic(p, d); }
  readRange(p: string, o: number, l: number) { return this.inner.readRange(p, o, l); }
  writeRange(p: string, o: number, d: Uint8Array) { return this.inner.writeRange(p, o, d); }
  async exists(p: string) {
    if (this.throwFor(p)) {
      const error = new Error(`a directory stands at ${p}`);
      error.name = "TypeMismatchError";
      throw error;
    }
    return this.inner.exists(p);
  }
  remove(p: string) { return this.inner.remove(p); }
  size(p: string) { return this.inner.size(p); }
  list(p: string) { return this.inner.list(p); }
  withLock<T>(s: string, fn: () => Promise<T>) { return this.inner.withLock(s, fn); }
}

function uniformSource(e: number, n: number): Uint8Array {
  const bytes = randomBytes(2 * (e + 32 * n));
  return new Uint8Array(bytes.buffer, bytes.byteOffset, bytes.byteLength);
}

async function browserPair(vfs: Vfs): Promise<string> {
  const res = await send(vfs, {
    op: "gen",
    label: "terminal-marker",
    sources: [{ name: "src.bin", declaredOrigin: "test material, operator-asserted", bytes: uniformSource(64, 8) }],
    encryptionBytes: 64,
    authRecords: 8,
    witnessClass: "browser-local-witness"
  });
  if (!res.ok || res.op !== "gen") {
    throw new Error(`gen failed: ${JSON.stringify(res)}`);
  }
  return res.pair.pairId;
}

describe("the §17 terminal marker fails closed in the browser engine too", { timeout: 120_000 }, () => {
  const MALFORMED: ReadonlyArray<{ name: string; bytes: string }> = [
    { name: "an empty file (a torn write)", bytes: "" },
    { name: "a truncated JSON object", bytes: '{"formatVersion":2,"pairId":"aaaa' },
    { name: "not JSON at all", bytes: "  not json" },
    { name: "JSON that is not an object", bytes: "[1,2,3]" },
    { name: "a JSON object naming a DIFFERENT pair", bytes: JSON.stringify({ formatVersion: 2, pairId: "f".repeat(32) }) }
  ];

  for (const shape of MALFORMED) {
    it(`refuses pair-destroyed when the tombstone is ${shape.name}`, async () => {
      const vfs = new MemoryVfs();
      const pairId = await browserPair(vfs);

      // Control first: this exact call succeeds without the marker.
      const before = await send(vfs, { op: "burn", pairId, as: "A", plaintext: new TextEncoder().encode("before") });
      expect(before.ok).toBe(true);

      await vfs.writeFileAtomic(`${pairId}/destroyed.json`, new TextEncoder().encode(shape.bytes));

      const after = await send(vfs, { op: "burn", pairId, as: "A", plaintext: new TextEncoder().encode("after") });
      expect(after.ok, `burn must refuse (${shape.name})`).toBe(false);
      expect((after as { reason?: string }).reason).toBe("pair-destroyed");
    });
  }

  it("an exists() that throws closes the boundary instead of reading as absent", async () => {
    const inner = new MemoryVfs();
    const pairId = await browserPair(inner);
    const vfs = new TypeMismatchOnExistsVfs(inner, (p) => p === `${pairId}/destroyed.json`);

    const burn = await send(vfs, { op: "burn", pairId, as: "A", plaintext: new TextEncoder().encode("after") });
    expect(burn.ok, "a tombstone whose presence cannot be decided must not permit a burn").toBe(false);

    const open = await send(vfs, { op: "status", pairId });
    expect(open.ok, "status must not proceed either").toBe(false);
  });
});

File: android/truepad-storage/src/test/kotlin/dev/systemslibrarian/truepad/storage/TerminalMarkerFailClosedTest.kt
package dev.systemslibrarian.truepad.storage

import dev.systemslibrarian.truepad.core.requiredSourceLength
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File
import java.nio.file.Files

/**
 * THE TERMINAL MARKER MUST FAIL CLOSED (FORMAT-V2.md §17).
 *
 * `destroyed.json` is the irreversible boundary: once it is durable the pair must
 * never perform a cryptographic operation again. So the gate asks "is this path
 * NOT KNOWN TO BE ABSENT", not "is there a readable regular file here" — every
 * way of being unreadable has to close the boundary, not open it.
 *
 * A REAL FAIL-OPEN, MEASURED RATHER THAN ASSUMED. `requireNotDestroyed` used
 * `Fs.exists`, which was `File.exists()`, which FOLLOWS SYMLINKS and answers
 * false for a symlink whose target is gone. A tombstone in that shape read as
 * absent and the destroyed pair became usable again — pad reuse. Node's
 * `existsSync` and Foundation's `fileExists` answer false there too, so the
 * Browser/CLI and iOS editions carried the identical defect and were corrected in
 * the same change; `tests/terminal-marker-fail-closed.test.ts` and
 * `TerminalMarkerFailClosedTests.swift` assert the same list of shapes.
 *
 * This is the sibling of [HandoffFailClosedTest]: the same "present but
 * unreadable is not absence" rule, applied to the marker that is permanent.
 *
 * LOSS IS ACCEPTABLE; REUSE IS NOT.
 */
class TerminalMarkerFailClosedTest {

    private fun tempRoot(prefix: String): File = Files.createTempDirectory(prefix).toFile()

    private fun freshPadOn(fs: Fs): String {
        val engine = fixedEngine(fs)
        val need = requiredSourceLength(256, 4).toInt()
        engine.gen(
            "terminal-fail-closed",
            listOf(SourceInput("s.bin", "declared", genBytes(need, 23))),
            256, 4, witnessKind = WitnessKind.LOCAL,
        )
        return FIXED_PAIR_ID
    }

    /** Replace the tombstone path with something that is NOT a readable file. */
    private fun plant(root: File, pairId: String, make: (File) -> Unit): Boolean {
        val marker = File(File(root, pairId), TOMBSTONE_FILE)
        marker.parentFile.mkdirs()
        if (Files.exists(marker.toPath(), java.nio.file.LinkOption.NOFOLLOW_LINKS)) {
            marker.deleteRecursively()
        }
        return try {
            make(marker)
            true
        } catch (_: Exception) {
            // Some filesystems disallow symlinks or special files. Skip rather
            // than fail — but never silently pass the assertion.
            false
        }
    }

    /** Every shape a tombstone can take that is NOT a well-formed readable file. */
    private val shapes: List<Pair<String, (File) -> Unit>> = listOf(
        "a symlink whose target does not exist" to { m: File ->
            Files.createSymbolicLink(m.toPath(), File(m.parentFile, "does-not-exist").toPath())
        },
        "a symlink to a deleted file" to { m: File ->
            val t = File(m.parentFile, "t.bin")
            t.writeText("x")
            Files.createSymbolicLink(m.toPath(), t.toPath())
            t.delete()
        },
        "a directory" to { m: File -> m.mkdirs() },
        "a non-empty directory" to { m: File ->
            m.mkdirs()
            File(m, "inner").writeText("x")
        },
        "an empty file (a torn write)" to { m: File -> m.writeBytes(ByteArray(0)) },
        "a truncated JSON object" to { m: File -> m.writeText("{\"formatVersion\":2,\"pairId\":\"aaaa") },
        "not JSON at all" to { m: File -> m.writeText("  not json") },
        "JSON that is not an object" to { m: File -> m.writeText("[1,2,3]") },
        "a JSON object naming a DIFFERENT pair" to { m: File ->
            m.writeText("{\"formatVersion\":2,\"pairId\":\"${"f".repeat(32)}\"}")
        },
        "a file with no read permission" to { m: File ->
            m.writeText("{}")
            check(m.setReadable(false, false)) { "could not drop read permission" }
        },
    )

    // ---- the gate ------------------------------------------------------------

    @Test
    fun everyUnreadableTombstoneShapeStillRefusesEveryConsumingVerb() {
        var exercised = 0
        for ((name, make) in shapes) {
            val root = tempRoot("tp-terminal")
            try {
                val fs = NioFs(root)
                val pairId = freshPadOn(fs)

                // CONTROL FIRST: without the marker this exact call succeeds, so
                // the refusal below is caused by the marker and nothing else.
                val engine = fixedEngine(fs)
                engine.burn(pairId, Party2.A, "before".toByteArray())

                if (!plant(root, pairId) { make(it) }) continue
                exercised += 1

                for (verb in listOf<Pair<String, () -> Unit>>(
                    "burn" to { engine.burn(pairId, Party2.A, "after".toByteArray()) },
                    "status" to { engine.status(pairId) },
                )) {
                    val refusal = refusalOf { verb.second() }
                    assertEquals(
                        "[$name] ${verb.first} must refuse pair-destroyed, not proceed",
                        "pair-destroyed", refusal.reason,
                    )
                }
            } finally {
                root.setWritable(true, false)
                root.walkBottomUp().forEach { it.setReadable(true, false); it.setWritable(true, false) }
                root.deleteRecursively()
            }
        }
        assertTrue("no shape was exercised — the suite proved nothing", exercised >= 8)
    }

    /**
     * The end-to-end shape of the original defect: destroy for real, then let a
     * restore or a sync tool leave a broken link where the tombstone was.
     */
    @Test
    fun aDestroyedPairWhoseTombstoneBecomesADanglingSymlinkStaysDestroyed() {
        val root = tempRoot("tp-terminal-e2e")
        try {
            val fs = NioFs(root)
            val pairId = freshPadOn(fs)
            val engine = fixedEngine(fs)
            engine.destroy(pairId, pairId, "operator destroy")

            val marker = File(File(root, pairId), TOMBSTONE_FILE)
            assertTrue("the tombstone should exist after destroy", marker.isFile)
            marker.delete()
            try {
                Files.createSymbolicLink(marker.toPath(), File(root, "gone-target").toPath())
            } catch (_: Exception) {
                return // no symlink support here; the shape test above still ran
            }
            assertFalse("precondition: File.exists() is exactly what fooled the old gate", marker.exists())

            assertEquals(
                "a destroyed pair must not resurrect because its tombstone became a broken link",
                "pair-destroyed",
                refusalOf { engine.burn(pairId, Party2.A, "resurrected?".toByteArray()) }.reason,
            )
        } finally {
            root.deleteRecursively()
        }
    }

    // ---- the Fs-level contract ------------------------------------------------

    /**
     * `Fs.exists` means NOT KNOWN TO BE ABSENT. Only a definitive "nothing is
     * here" may be false.
     */
    @Test
    fun existsReportsPresenceForEveryNonAbsentShapeAndAbsenceOnlyWhenDefinite() {
        val root = tempRoot("tp-terminal-fs")
        try {
            val fs = NioFs(root)
            var exercised = 0
            for ((index, shape) in shapes.withIndex()) {
                val probe = File(root, "probe-$index")
                try {
                    shape.second(probe)
                } catch (_: Exception) {
                    continue
                }
                exercised += 1
                assertTrue("[${shape.first}] must read as present", fs.exists("probe-$index"))
            }
            assertTrue("no shape was exercised", exercised >= 8)

            // The ONE definitive negative: there is no such path.
            assertFalse(fs.exists("nothing-here"))
            assertFalse(fs.exists("no/such/directory/at/all.json"))

            // And a path whose parent is not a directory is NOT a definitive
            // negative -- the JDK does not map that errno portably, so the
            // conservative answer is the one all three editions give.
            File(root, "plain.bin").writeText("x")
            assertTrue(
                "a path that cannot be resolved is not a path known to be clear",
                fs.exists("plain.bin/under"),
            )
        } finally {
            root.walkBottomUp().forEach { it.setReadable(true, false); it.setWritable(true, false) }
            root.deleteRecursively()
        }
    }
}
